lib/vtls/openssl.c

```diff
@@ -5364,31 +5364,32 @@ static ssize_t ossl_recv(struct Curl_cfilter *cf,
         nread = -1;
         goto out;
       }
-      /* For debug builds be a little stricter and error on any
-         SSL_ERROR_SYSCALL. For example a server may have closed the connection
-         abruptly without a close_notify alert. For compatibility with older
-         peers we do not do this by default. #4624
-
-         We can use this to gauge how many users may be affected, and
-         if it goes ok eventually transition to allow in dev and release with
-         the newest OpenSSL: #if (OPENSSL_VERSION_NUMBER >= 0x10101000L) */
-#ifdef DEBUGBUILD
-      if(err == SSL_ERROR_SYSCALL) {
-        int sockerr = SOCKERRNO;
-        if(sockerr)
-          Curl_strerror(sockerr, error_buffer, sizeof(error_buffer));
+      else if(err == SSL_ERROR_SYSCALL) {
```

Comment:
```suggestion
#if OPENSSL_VERSION_NUMBER >= 0x30000000L
      else if(err == SSL_ERROR_SYSCALL) {
```

This could cause problems for legacy versions, what do you think about limit OpenSSL 3

Reply:
Since we use a `BIO` nowadays, do you think that old problem is still relevant? The reporter tested OpenSSL 1.1.1 and verified it works for them. So legacy would be some 0.98 zombie?

Reply:
#17471 as reported by @mkauf appears to have been solved by f1e1c8b and the behavior with OpenSSL 1.1.1 is AFAICT unchanged (as in with that OpenSSL version both f1e1c8b and `f1e1c8b~1` there's no libcurl error for the missing terminator for non-debug builds).

Here your changes will make it so that OpenSSL 1.1.1 (or anything before 3.x I think) libcurl will return CURLE_RECV_ERROR if there's a missing terminator. I tested in OpenSSL 1.1.1t with this branch and curl returns `curl: (56) Connection closed abruptly` for the missing terminator. What I am saying is that could cause problems for users of those old versions that are more likely to be connecting to old servers that may do that.

~~~diff
diff --git a/lib/vtls/openssl.c b/lib/vtls/openssl.c
index 6d02b00..1f4e632 100644
--- a/lib/vtls/openssl.c
+++ b/lib/vtls/openssl.c
@@ -5364,6 +5364,9 @@ static ssize_t ossl_recv(struct Curl_cfilter *cf,
         nread = -1;
         goto out;
       }
+/* Error on abrupt connection closures except for legacy versions of OpenSSL
+   since those versions may be connecting to older peers with that behavior. */
+#if defined(DEBUGBUILD) || (OPENSSL_VERSION_NUMBER >= 0x30000000L)
       else if(err == SSL_ERROR_SYSCALL) {
         if(octx->io_result) {
           /* logging handling in underlying filter already */
@@ -5390,6 +5393,7 @@ static ssize_t ossl_recv(struct Curl_cfilter *cf,
         nread = -1;
         goto out;
       }
+#endif
     }
   }
 
~~~

Reply:
Maybe I am confused here. When I run curl + openssl 1.1.1 in this PR, the `test_05_04` for HTTP/1.0 is a HTTP/1.0 server reply without `Content-Length:` that requires a TSL close notify to succeed. 

Since the server in this test does not send it, curl is expected to detect that, and it does now. Before this PR, it only did that in debug builds. The local OpenSSL version has no influence.

If I understand you correctly, you propose to suppress this check for *new* curl builds with *old* OpenSSL versions, because they are *more likely* to contact old servers that are broken.

I think that gives - to the user - surprising behaviour when a server works on one curl 8.15.0 and fails with another curl 8.15.0, just because the linked OpenSSL version differs.

For curl to work with such servers, we need to add something like `--ssl-ignore-unexpected-eof`.

Reply:
I agree with @icing. I don't think building with OpenSSL 1 should be a reason for curl to act differently.

I think we should go with *do not silently accept unexpected EOF* as this PR proposes and only if we are pushed into a corner from users who really cannot survive without this ability, consider a version of an *ignore-unexpected-eof* option.